gpbft/emulator/emulator.go

```diff
@@ -0,0 +1,93 @@
+package emulator
+
+import (
+	"fmt"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+)
+
+// Emulator emulates the core GPBFT protocol for one honest participant, and
+// allows frame-by-frame control over progress of a GPBFT instance.
+//
+// Progress may be made by calling Emulator.ReceiveMessage,
+// Emulator.ReceiveBroadcast or Emulator.ReceiveAlarm.
+type Emulator struct {
+	subject *gpbft.Participant
+	host    *host
+}
+
+// New instantiates a new Emulator with the given GPBFT options. See
+// Emulator.Start.
+func New(o ...gpbft.Option) (*Emulator, error) {
+	h := newHost()
+	participant, err := gpbft.NewParticipant(h, o...)
+	if err != nil {
+		return nil, err
+	}
+	return &Emulator{
+		subject: participant,
+		host:    h,
+	}, nil
+}
+
+// Start starts emulation for the given instance and signals the start of
+// instance to the emulated honest gpbft.Participant.
+//
+// See NewInstance.
+func (e *Emulator) Start(instance *Instance) error {
+	if err := e.host.setInstance(instance); err != nil {
+		return err
+	}
+	return e.subject.StartInstance(instance.id)
+}
+
+// ReceiveAlarm triggers any pending alarms. Returns true if the alarm is
+// triggered along with any errors that may occur as a result of triggering the
+// alarm. Otherwise returns false with no error.
+func (e *Emulator) ReceiveAlarm() (bool, error) {
+	if e.host.maybeTriggerAlarm() {
+		return true, e.subject.ReceiveAlarm()
+	}
+	return false, nil
+}
+
+// ReceiveMessage delivers a message to the emulated participant from the given
+// sender. The sender must not be the ID of the honest participant as
+// impersonation of messages to the participant is not supported.
+//
+// Returns the final built and delivered message along with any errors that may
+// occur as a result of message delivery. If the building of the message fails
+// nil message is returned with error.
+func (e *Emulator) ReceiveMessage(from gpbft.ActorID, mb *gpbft.MessageBuilder) (*gpbft.GMessage, error) {
+	if from == e.host.id {
+		return nil, fmt.Errorf("subject impersonation is not supported")
+	}
+	mb.SetNetworkName(e.host.NetworkName())
+	mb.SetSigningMarshaler(e.host.adhocSigning)
+	msg, err := mb.Build(e.host.adhocSigning, from)
```

Comment:
I was expecting to make assertions directly on the thing the participant sent to us: the MessageBuilder. Initially I was confused why this is being built into a message first. I since realised it's for delivery of messages back to the participant. If you didn't do this, I think you could remove even the CRC signing and replace the host signature verification calls with simple mocks that just return true/false (and maybe check arguments, for one or two tests).